File: Yandex/Practicum/Algorightms_and_Data_Structures/7_sprint/J._Путешествие/solution.cpp
#include <iostream>
#include <vector>


void solution() {
    /*
    Time Complexity O(N^2)
    Memory Complexity O(N)
    */
    int n;
    std::scanf("%d", &n);
    std::vector<int> nums (n, 0);
    for (int i = 0; i < n; ++i) {
        std::cin >> nums[i];
    }
    std::vector<int> dp (n + 1, 0);
    std::vector<std::vector<int>> paths (n + 1, std::vector<int>());
    for (int i = 1; i <= n; ++i) {
        for (int j = 0; j < i; ++j) {
            if ((nums[j - 1] < nums[i - 1]) && (dp[i] < dp[j])) {
                dp[i] = dp[j];
                paths[i] = paths[j];
            }
        }
        ++dp[i];
        paths[i].push_back(i);
    }
    int lisIdx = 0;
    for (int idx = 1; idx <= n; ++idx) {
        if (dp[idx] > dp[lisIdx]) lisIdx = idx;
    }
    std::printf("%d\n", dp[lisIdx]);
    if (dp[lisIdx]) {
        for (int& num : paths[lisIdx]) std::cout << num << " ";
        std::cout << "\n";
    }
}


int main() {
    solution();
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/7_sprint/J._Путешествие/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N^2)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    nums: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    dp: list[int] = [0] * (n + 1)
    paths: list[list[int]] = [[] for _ in range(n + 1)]
    for i in range(1, n + 1):
        for j in range(i):
            if nums[j - 1] < nums[i - 1] and dp[i] < dp[j]:
                dp[i] = dp[j]
                paths[i] = paths[j].copy()
        dp[i] += 1
        paths[i].append(i)
    lis_idx: int = 0
    for idx in range(1, n + 1):
        if dp[idx] > dp[lis_idx]: lis_idx = idx
    sys.stdout.write(str(dp[lis_idx]) + '\n')
    if dp[lis_idx]:
        sys.stdout.write(' '.join(str(idx) for idx in paths[lis_idx]) + '\n')


if __name__ == '__main__':
    solution()